hash/bucketer.go

```diff
@@ -0,0 +1,93 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// This file contains the utilities to make bucketing decisions.
+
+package hash
+
+import (
+	"sync"
+
+	lru "github.com/hashicorp/golang-lru"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+// BucketSet answers to what bucket does key X belong in a
+// consistent manner (consistent as in consistent hashing).
+// In addition Bucket implements reconciler.Bucket interface, so it
+// can be used both in leader election and in routing applications.
+type BucketSet struct {
```

Comment:
Personally I'd prefer to split bucket interface and buckeset into two structs because bucketset sounds like a set of multiple buckets. Maybe we can give it a better name?

Reply:
Not sure... it kind of takes care of multiple buckets as well.

Reply:
My understanding was that we were talking about a BucketSet interface to complement and aggregate multiple Buckets lead by the same replica.  The intention being to avoid the loop over the buckets a replica leads each time we enqueue.

Also, must this be public?  😉   Why not return the interface?

Reply:
Besides go philosophy being return types accept interfaces, we have Owner method.